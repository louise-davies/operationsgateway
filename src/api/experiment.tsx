import { useQuery, UseQueryResult } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { ExperimentParams } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const fetchExperiment = (apiUrl: string): Promise<ExperimentParams[]> => {
  return axios
    .get(`${apiUrl}/experiments`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useExperiment = (): UseQueryResult<
  ExperimentParams[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['experiments'],

    queryFn: () => fetchExperiment(apiUrl),
  });
};
