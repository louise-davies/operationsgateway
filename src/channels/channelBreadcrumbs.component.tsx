import { NavigateNext } from '@mui/icons-material';
import { Breadcrumbs, Link, Typography } from '@mui/material';
import React from 'react';

type ChannelBreadcrumbsProps = {
  currNode: string;
  setCurrNode: (newNode: string) => void;
};

const ChannelBreadcrumbs = (props: ChannelBreadcrumbsProps) => {
  const { currNode, setCurrNode } = props;
  // special handling for '/' otherwise we get ['','']
  const nodePath = currNode === '/' ? [''] : currNode.split('/');

  return (
    <Breadcrumbs
      separator={<NavigateNext fontSize="small" />}
      aria-label="breadcrumb"
    >
      {nodePath.map((value, currIndex) => {
        const label = value.length ? value : 'All Channels';
        // final node, render as non-"link"
        if (currIndex === nodePath.length - 1) {
          return (
            <Typography key={label} color="text.primary">
              {label}
            </Typography>
          );
        } else {
          return (
            <Link
              key={label}
              href="#"
              onClick={(ev) => {
                // prevent default to stop the href="#" adding a hash to the URL
                // and potentially jumping the page
                ev.preventDefault();
                setCurrNode(nodePath.slice(0, currIndex + 1).join('/'));
              }}
            >
              {label}
            </Link>
          );
        }
      })}
    </Breadcrumbs>
  );
};

export default ChannelBreadcrumbs;
