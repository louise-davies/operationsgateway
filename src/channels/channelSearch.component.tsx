import { Search } from '@mui/icons-material';
import {
  Autocomplete,
  type FilterOptionsState,
  InputAdornment,
  TextField,
} from '@mui/material';
import React from 'react';
import { FullChannelMetadata } from '../app.types';

type ChannelSearchProps = {
  currPathAndChannel: string;
  channels: FullChannelMetadata[];
  onSearchChange: (channel: FullChannelMetadata) => void;
};

const filterOptions = (
  options: FullChannelMetadata[],
  { inputValue }: FilterOptionsState<FullChannelMetadata>
) => {
  const searchString = inputValue.toLowerCase();

  return options.filter((option) => {
    const name = option?.name?.toLowerCase();
    const systemName = option.systemName.toLowerCase();
    // need to search both friendly and system names for a match
    return name?.includes(searchString) || systemName.includes(searchString);
  });
};

const ChannelSearch = (props: ChannelSearchProps) => {
  const { onSearchChange, channels, currPathAndChannel } = props;

  const [value, setValue] = React.useState<FullChannelMetadata | null>(null);
  const [inputValue, setInputValue] = React.useState('');

  React.useEffect(() => {
    // clear the search box if the user clicks another channel / navigates via breadcrumbs
    if (value && currPathAndChannel !== `${value.path}/${value.systemName}`) {
      setValue(null);
      setInputValue('');
    }
  }, [currPathAndChannel, value]);

  return (
    <Autocomplete
      fullWidth
      filterOptions={filterOptions}
      value={value}
      inputValue={inputValue}
      onInputChange={(event, newInputValue) => setInputValue(newInputValue)}
      size="small"
      options={channels}
      getOptionLabel={(option) =>
        option.name
          ? `${option.name} (${option.systemName})`
          : option.systemName
      }
      blurOnSelect
      onChange={(event: unknown, newValue: FullChannelMetadata | null) => {
        if (newValue) onSearchChange(newValue);
        setValue(newValue);
      }}
      renderInput={(params) => (
        <TextField
          {...params}
          label="Search data channels"
          InputProps={{
            ...params.InputProps,
            startAdornment: (
              <InputAdornment position="start">
                <Search />
              </InputAdornment>
            ),
          }}
        />
      )}
    />
  );
};

export default ChannelSearch;
