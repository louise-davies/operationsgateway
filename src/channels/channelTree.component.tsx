import {
  Checkbox,
  List,
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
} from '@mui/material';
import React from 'react';
import { TreeNode } from './channelsDialogue.component';

type ChannelTreeProps = {
  currNode: string;
  tree: TreeNode;
  setCurrNode: (newNode: string) => void;
  handleChannelChecked: (channel: string, checked: boolean) => void;
};

const ChannelTree = (props: ChannelTreeProps) => {
  const { currNode, tree, setCurrNode, handleChannelChecked } = props;

  const nodes = currNode
    .split('/')
    .filter((el) => el)
    .reduce((prev, curr) => prev.children?.[curr] as TreeNode, tree);

  const leaf = !('children' in nodes.children[Object.keys(nodes.children)[0]]);

  return (
    <List dense>
      {Object.keys(nodes.children).map((value) => {
        const labelId = `checkbox-list-label-${value}`;
        return (
          <ListItem key={value} disablePadding disableGutters>
            <ListItemButton
              onClick={() => {
                if (!leaf) {
                  setCurrNode(`${currNode !== '/' ? currNode : ''}/${value}`);
                } else {
                  if (value !== 'timestamp')
                    handleChannelChecked(value, nodes.children[value].checked);
                  // add to channels? open up side panel?
                }
              }}
            >
              <ListItemIcon>
                <Checkbox
                  checked={nodes.children[value].checked}
                  tabIndex={-1}
                  disableRipple
                  disabled={!leaf || value === 'timestamp'}
                  size="small"
                  inputProps={{ 'aria-labelledby': labelId }}
                />
              </ListItemIcon>
              <ListItemText id={labelId} primary={value} />
            </ListItemButton>
          </ListItem>
        );
      })}
    </List>
  );
};

export default ChannelTree;
