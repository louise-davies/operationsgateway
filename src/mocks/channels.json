{
  "_id": "123",
  "channels": {
    "CHANNEL_ABCDE": {
      "name": "Channel_ABCDE",
      "path": "/Channels/1/ABCDE",
      "type": "scalar"
    },
    "CHANNEL_BCDEF": {
      "name": "Channel_BCDEF",
      "path": "/Channels/1/BCDEF",
      "type": "image"
    },
    "CHANNEL_CDEFG": {
      "name": "Channel_CDEFG",
      "path": "/Channels/1/CDEFG",
      "type": "waveform",
      "x_units": "X",
      "y_units": "Y"
    },
    "CHANNEL_DEFGH": {
      "name": "Channel_DEFGH",
      "path": "/Channels/2/DEFGH",
      "type": "scalar",
      "notation": "scientific",
      "precision": 1
    },
    "CHANNEL_EFGHI": {
      "name": "Channel_EFGHI",
      "path": "/Channels/2/EFGHI",
      "type": "image"
    },
    "CHANNEL_FGHIJ": {
      "name": "Channel_FGHIJ",
      "path": "/Channels/2/FGHIJ",
      "type": "waveform"
    },
    "CHANNEL_GHIJK": {
      "name": "Channel_GHIJK",
      "path": "/Channels/3/GHIJK",
      "type": "scalar"
    }
  }
}
