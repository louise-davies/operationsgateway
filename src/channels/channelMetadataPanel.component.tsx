import AddIcon from '@mui/icons-material/Add';
import RemoveIcon from '@mui/icons-material/Remove';
import {
  Button,
  Divider,
  Stack,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from '@mui/material';
import React from 'react';
import { useChannelSummary } from '../api/channels';
import {
  FullChannelMetadata,
  isChannelMetadataScalar,
  isChannelMetadataWaveform,
} from '../app.types';
import {
  TraceOrImageThumbnail,
  renderTimestamp,
} from '../table/cellRenderers/cellContentRenderers';

type ChannelMetadataPanelProps = {
  displayedChannel: FullChannelMetadata | undefined;
  isChannelSelected: boolean;
  onSelectChannel: (channel: string) => void;
  onDeselectChannel: (channel: string) => void;
};

const Heading = (props: React.ComponentProps<typeof Typography>) => {
  const { children, ref, ...restProps } = props;
  return (
    <Typography
      variant="body1"
      component="h3"
      gutterBottom
      sx={{ fontWeight: 'bold' }}
      {...restProps}
    >
      {children}
    </Typography>
  );
};
const Body = (props: React.ComponentProps<typeof Typography>) => {
  const { children, ...restProps } = props;
  return (
    <Typography variant="body2" gutterBottom {...restProps}>
      {props.children}
    </Typography>
  );
};

const ChannelMetadataPanel = (props: ChannelMetadataPanelProps) => {
  const {
    displayedChannel,
    isChannelSelected,
    onSelectChannel,
    onDeselectChannel,
  } = props;

  const { data: channelSummary } = useChannelSummary(
    displayedChannel?.systemName
  );

  function addCurrentChannel() {
    if (displayedChannel?.systemName) {
      onSelectChannel(displayedChannel.systemName);
    }
  }

  function removeCurrentChannel() {
    if (displayedChannel?.systemName) {
      onDeselectChannel(displayedChannel.systemName);
    }
  }

  if (displayedChannel) {
    return (
      <>
        <Stack
          direction="row"
          justifyContent="space-between"
          alignItems="center"
        >
          <Heading gutterBottom={false}>
            {displayedChannel?.name ?? displayedChannel.systemName}
          </Heading>
          {isChannelSelected ? (
            <Button
              variant="outlined"
              startIcon={<RemoveIcon />}
              size="small"
              onClick={removeCurrentChannel}
            >
              Remove this channel
            </Button>
          ) : (
            <Button
              variant="contained"
              startIcon={<AddIcon />}
              size="small"
              onClick={addCurrentChannel}
            >
              Add this channel
            </Button>
          )}
        </Stack>
        {displayedChannel?.name && (
          <Body>System name: {displayedChannel.systemName}</Body>
        )}
        {displayedChannel?.description && (
          <Body>{displayedChannel.description}</Body>
        )}
        <Body>Channel type: {displayedChannel.type}</Body>
        {isChannelMetadataScalar(displayedChannel) &&
          displayedChannel.units && (
            <Body>Units: {displayedChannel.units}</Body>
          )}
        {isChannelMetadataWaveform(displayedChannel) && (
          <>
            <Body gutterBottom={false}>
              X Units: {displayedChannel.x_units}
            </Body>
            <Body>Y Units: {displayedChannel.y_units}</Body>
          </>
        )}
        {displayedChannel?.historical && (
          <Body fontWeight="bold">This channel is historical</Body>
        )}
        {channelSummary && (
          <>
            <Divider />
            <Typography
              variant="body2"
              component="h4"
              gutterBottom
              sx={{ fontWeight: 'bold', paddingTop: 1 }}
            >
              Data Summary
            </Typography>
            <Body gutterBottom={false}>
              First data date: {renderTimestamp(channelSummary.first_date)}
            </Body>
            <Body>
              Most recent data date:{' '}
              {renderTimestamp(channelSummary.most_recent_date)}
            </Body>
            <Typography
              variant="body2"
              component="h5"
              gutterBottom
              sx={{ fontWeight: 'bold', paddingTop: 1 }}
            >
              Recent Data
            </Typography>
            <TableContainer>
              <Table aria-label="recent data" size="small">
                <TableHead>
                  <TableRow>
                    <TableCell>Time</TableCell>
                    <TableCell>Data</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {channelSummary.recent_sample.map((sample) => {
                    const [timestamp, data] = Object.entries(sample)[0];
                    const formattedTimestamp = renderTimestamp(timestamp);
                    return (
                      <TableRow
                        key={formattedTimestamp}
                        sx={{
                          '&:last-child td, &:last-child th': { border: 0 },
                        }}
                      >
                        <TableCell component="th" scope="row">
                          {formattedTimestamp}
                        </TableCell>
                        <TableCell>
                          {isChannelMetadataScalar(displayedChannel) ? (
                            data
                          ) : (
                            <TraceOrImageThumbnail
                              base64Data={data as string}
                              alt={`${
                                displayedChannel?.name ??
                                displayedChannel.systemName
                              } ${
                                displayedChannel.type
                              } at ${formattedTimestamp}`}
                            />
                          )}
                        </TableCell>
                      </TableRow>
                    );
                  })}
                </TableBody>
              </Table>
            </TableContainer>
          </>
        )}
      </>
    );
  } else {
    return (
      <>
        <Heading>Data Channels Help</Heading>
        <Body>
          There are two options to locate a channel. You can either search or
          use the navigation on the left.
        </Body>
        <Body>
          Channel metadata can also be viewed. This can be displayed by clicking
          a channel name
        </Body>
      </>
    );
  }
};

export default ChannelMetadataPanel;
