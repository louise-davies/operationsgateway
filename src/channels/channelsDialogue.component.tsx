import React from 'react';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Divider,
  Grid,
  InputAdornment,
  TextField,
} from '@mui/material';
import { Search } from '@mui/icons-material';
import { useChannels } from '../api/channels';
import { FullChannelMetadata } from '../app.types';
import ChannelTree from './channelTree.component';
import ChannelBreadcrumbs from './channelBreadcrumbs.component';
import { useAppDispatch, useAppSelector } from '../state/hooks';
import {
  selectSelectedIdsIgnoreOrder,
  updateSelectedColumns,
} from '../state/slices/tableSlice';
import { createSelector } from '@reduxjs/toolkit';

interface ChannelsDialogueProps {
  open: boolean;
  onClose: () => void;
}

export interface TreeNode {
  name: string;
  checked: boolean;
  children: Record<
    string,
    TreeNode | (FullChannelMetadata & { checked: boolean })
  >;
}

/**
 * @returns A selector for a tree representing the channel hierarchy,
 * which is used by components in the channels folder
 * @params state - redux state - isn't actually used! (means we get nice memoisation from reselect)
 * @params availableChannels - array of all the channels the user can select
 * @params selectedIds - array of all the channels currently selected
 */
export const selectChannelTree = createSelector(
  (
    state: unknown,
    availableChannels: FullChannelMetadata[],
    selectedIds: string[]
  ) => availableChannels,
  (
    state: unknown,
    availableChannels: FullChannelMetadata[],
    selectedIds: string[]
  ) => selectedIds,
  (availableChannels, selectedIds) => {
    const tree: TreeNode = { name: '/', children: {}, checked: false };
    availableChannels.forEach((channel) => {
      const { path } = channel;
      // split the path and remove any empty strings
      const splitPath = path.split('/').filter((el) => el);
      splitPath.reduce((prev, curr, currIndex) => {
        // init treenode if not already initialised
        if (!prev.children?.[curr]) {
          prev.children[curr] = {
            name: curr,
            checked: false,
            children: {},
          };
        }
        const node = prev.children[curr] as TreeNode;
        // last item, so therefore is the channel
        if (currIndex === splitPath.length - 1) {
          node.children[channel.systemName] = {
            ...channel,
            checked: selectedIds.includes(channel.systemName),
          };
        }
        // return the child node, so we drill deeper into the tree
        return node;
      }, tree);
    });

    return tree;
  }
);

const ChannelsDialogue = (props: ChannelsDialogueProps) => {
  const { open, onClose } = props;

  const { data: channels } = useChannels();

  const appliedSelectedIds = useAppSelector((state) =>
    selectSelectedIdsIgnoreOrder(state)
  );

  const [selectedIds, setSelectedIds] = React.useState(appliedSelectedIds);

  React.useEffect(() => {
    setSelectedIds(appliedSelectedIds);
  }, [appliedSelectedIds]);

  const channelTree = useAppSelector((state) =>
    selectChannelTree(state, channels ?? [], selectedIds)
  );

  const dispatch = useAppDispatch();

  const onChannelSelect = React.useCallback((channel: string): void => {
    setSelectedIds((selectedIds) => [...selectedIds, channel]);
  }, []);

  const onChannelDeselect = React.useCallback((channel: string): void => {
    setSelectedIds((selectedIds) =>
      selectedIds.filter((colId) => colId !== channel)
    );
  }, []);

  const handleChannelChecked = React.useCallback(
    (channel: string, checked: boolean) => {
      checked ? onChannelDeselect(channel) : onChannelSelect(channel);
    },
    [onChannelDeselect, onChannelSelect]
  );

  const [currNode, setCurrNode] = React.useState('/');

  return (
    <Dialog open={open} onClose={onClose} maxWidth="lg" fullWidth>
      <Grid container columnSpacing={2} alignItems="center">
        <Grid item xs>
          <DialogTitle>Data Channels</DialogTitle>
        </Grid>
        <Grid item xs>
          <TextField
            size="small"
            label="Search data channels"
            fullWidth
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  <Search />
                </InputAdornment>
              ),
            }}
          />
        </Grid>
      </Grid>
      <DialogContent
        sx={{
          border: 'thin lightgrey',
          borderStyle: 'solid none',
          paddingTop: '8px !important',
          paddingBottom: '8px',
          overflowY: 'unset',
        }}
      >
        <ChannelBreadcrumbs currNode={currNode} setCurrNode={setCurrNode} />
      </DialogContent>
      <DialogContent>
        <Grid container columnSpacing={2}>
          <Grid item xs>
            <ChannelTree
              currNode={currNode}
              tree={channelTree ?? { name: '/', children: {}, checked: false }}
              setCurrNode={setCurrNode}
              handleChannelChecked={handleChannelChecked}
            />
          </Grid>
          <Divider orientation="vertical" flexItem />
          <Grid item xs>
            Help / info
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Close</Button>
        <Button
          onClick={() => {
            dispatch(updateSelectedColumns(selectedIds));
            onClose();
          }}
        >
          Add Channels
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default ChannelsDialogue;
