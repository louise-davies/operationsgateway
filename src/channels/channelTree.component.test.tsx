import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { staticChannels } from '../api/channels';
import { FullChannelMetadata } from '../app.types';
import channelsJson from '../mocks/channels.json';
import { TreeNode } from './channelsDialogue.component';
import ChannelTree from './channelTree.component';

describe('Channel Tree', () => {
  let currNode = '';
  const setCurrNode = jest.fn();
  const handleChannelChecked = jest.fn();
  const handleChannelSelected = jest.fn();
  const channel1 = {
    ...Object.values(channelsJson.channels)[0],
    systemName: Object.keys(channelsJson.channels)[0],
  } as FullChannelMetadata;
  const channel2 = {
    ...Object.values(channelsJson.channels)[1],
    systemName: Object.keys(channelsJson.channels)[1],
  } as FullChannelMetadata;

  const tree: TreeNode = {
    name: '/',
    checked: false,
    children: {
      test_1: {
        name: 'test_1',
        checked: undefined,
        children: {
          timestamp: { ...staticChannels['timestamp'], checked: false },
          [channel1.systemName]: {
            ...channel1,
            checked: false,
          },
          [channel2.systemName]: {
            ...channel2,
            checked: true,
          },
        },
      },
      test_2: {
        name: 'test_2',
        checked: false,
        children: {
          test_3: { name: 'test_3', checked: false, children: {} },
        },
      },
    },
  };
  const createView = () => {
    return render(
      <ChannelTree
        currNode={currNode}
        setCurrNode={setCurrNode}
        handleChannelChecked={handleChannelChecked}
        handleChannelSelected={handleChannelSelected}
        tree={tree}
      />
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly for root', () => {
    currNode = '/';
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render correctly for sub node', () => {
    currNode = '/test_1';
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should call setCurrNode when non-leaf node is clicked', async () => {
    const user = userEvent.setup();
    currNode = '/';
    createView();

    await user.click(screen.getByText('test_2'));
    expect(setCurrNode).toHaveBeenCalledWith('/test_2');
  });

  it('should call setCurrNode when non-leaf node is clicked from non-leaf node', async () => {
    const user = userEvent.setup();
    currNode = '/test_2';
    createView();

    await user.click(screen.getByText('test_3'));
    expect(setCurrNode).toHaveBeenCalledWith('/test_2/test_3');
  });

  it('should call handleChannelChecked when leaf node checkbox is clicked', async () => {
    const user = userEvent.setup();
    currNode = '/test_1';
    createView();

    await user.click(
      screen.getByRole('checkbox', {
        name: channel1.name,
      })
    );
    expect(handleChannelChecked).toHaveBeenCalledWith(
      channel1.systemName,
      false
    );

    handleChannelChecked.mockClear();

    await user.click(
      screen.getByRole('checkbox', {
        name: channel2.name,
      })
    );
    expect(handleChannelChecked).toHaveBeenCalledWith(
      channel2.systemName,
      true
    );
  });

  it('should call handleChannelSelected when leaf node is clicked', async () => {
    const user = userEvent.setup();
    currNode = '/test_1';
    createView();

    await user.click(
      screen.getByLabelText(channel1.name ?? channel1.systemName)
    );
    expect(handleChannelSelected).toHaveBeenCalledWith({
      ...channel1,
      checked: false,
    });

    handleChannelChecked.mockClear();

    await user.click(
      screen.getByLabelText(channel2.name ?? channel2.systemName)
    );
    expect(handleChannelSelected).toHaveBeenCalledWith({
      ...channel2,
      checked: true,
    });
  });
});
