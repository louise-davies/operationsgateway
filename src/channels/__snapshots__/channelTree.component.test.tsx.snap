// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel Tree should render correctly for root 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            aria-disabled="true"
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
            tabindex="-1"
          >
            <input
              aria-labelledby="checkbox-list-label-test_1"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              disabled=""
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-test_1"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            test_1
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            aria-disabled="true"
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
            tabindex="-1"
          >
            <input
              aria-labelledby="checkbox-list-label-test_2"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              disabled=""
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-test_2"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            test_2
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Channel Tree should render correctly for sub node 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            aria-disabled="true"
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
            tabindex="-1"
          >
            <input
              aria-labelledby="checkbox-list-label-timestamp"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              disabled=""
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-timestamp"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            Time
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              aria-labelledby="checkbox-list-label-test_1"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-test_1"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            Test 1
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters css-178cqt3-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-355xbt-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              aria-labelledby="checkbox-list-label-test_2"
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-test_2"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            test_2
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
